lychee-lib/src/client.rs

```diff
@@ -454,6 +472,21 @@ impl Client {
         Ok(Response::new(uri.clone(), status, source))
     }
 
+    /// Returns optional basic auth based on if the URI matches any of the
+    /// selectors.
+    pub fn select_basic_auth_by_uri<'a>(
+        &'a self,
+        uri: &'a Uri,
+    ) -> Option<&'a Authorization<Basic>> {
+        for selector in &self.basic_auth {
```

Comment:
It's not a big deal, but that's O(n). If there's a lot of links or auth headers, this could slow down the checking process. We could consider using a [`RegexSet`](https://docs.rs/regex/latest/regex/struct.RegexSet.html) to achieve better performance.

Reply:
I moved the complete URI matching to the link collection phase. Now we do the heavy work upfront and not when making the requests. See 1083855